### ISSUE TITLE
HFS broken in 4.8

### ISSUE THREAD

**Issue by jonstewart** [sleuthkit/sleuthkit#1821]
The condition in this commit is incorrect: https://github.com/sleuthkit/sleuthkit/commit/218a868e1f273a89a3ddd9b36b63ef381915bac0

Patch and more issue details are forthcoming in next 24 hours. We have not been able to process HFS in 4.8 with this commit.

**Comment by uckelman-sf** [comment]
There are three problems preventing correct HFS/HFS+ processing in 4.8.0:

1. 218a868 and f9e8cbc0 introduce checks to prevent reading beyond the end of allocated memory, which were introduced in in response to #1076, in which Google fuzzing found two memory errors. Whilst the two commits do prevent invalid reads, they also cause HFS catalog records near the ends of nodes to be skipped during filesystem walks.

HFS catalog records are variable length. The checks introduced in f9e8cbc0 ("Fix for heap OOB memory read issue 1076") and 218a868e ("Check that the allocated node is large enough to be treated as a hfs_btree_key_cat.") incorrectly assume that all catalog records have the maximum possible length, and thus halt processing of a node when there is still sufficient space for catalog records which are shorter. The correct checks in PR #1822 compare the remaining space in the node with the actual declared length of the record to be read.

The invalid reads found in #1076 can be checked in valgrind using the test image there as follows: `valgrind bin/fls -b 512 -f hfs poc-96b218b5aee7dfd066ae1ff9bd52833e6604d9e5acb0dd97052f45a418b1c16_min`
I have verified that (1) using 4.7.0, the invalid reads do happen, (2) using 4.8.0, the invalid reads do not happen, and (3) using 4.8.0 + PR #1822 the invalid reads also do not happen.

2. `tsk_fs_type_to_name()` stopped being able to produce a name string for HFS or HFS+ in 4.8.0 due to a change in the type enum when APFS support was added, so anything that relied on getting a name string for HFS or HFS+ will now fail. This is fixed in PR #1801.

3. In 4.8.0, `zlib_inflate()` started being used in more than one source file, but no function prototype was ever added for it, which causes linking to fail in some circumstances. PR #1815 fixes this.
